File: src/56_medium_MergeIntervals.py
#!/usr/bin/python3

'''
Given an array of intervals where intervals[i] = [starti, endi], merge all overlapping intervals, and return an array of the non-overlapping intervals that cover all the intervals in the input.

 

Example 1:

Input: intervals = [[1,3],[2,6],[8,10],[15,18]]
Output: [[1,6],[8,10],[15,18]]
Explanation: Since intervals [1,3] and [2,6] overlaps, merge them into [1,6].
Example 2:

Input: intervals = [[1,4],[4,5]]
Output: [[1,5]]
Explanation: Intervals [1,4] and [4,5] are considered overlapping.
'''

def merge(intervals):
    if (len(intervals) == 0):
        return list()
    
    # key is first item, value is the largest second item 
    beg2end = dict()
    
    for item in intervals:
        if beg2end.get(item[0]) is None:
            beg2end[item[0]] = item[1]
        else:
            if beg2end[item[0]] < item[1]:
                beg2end[item[0]] = item[1]
    
    keys = sorted(beg2end.keys())
    
    result = list()
    index = 0
    curr_beg, curr_end = keys[0], beg2end[keys[0]]
    while index < len(keys):
        if (index == len(keys)-1):
            result.append([keys[-1], beg2end[keys[-1]]])
            break 
    
        b2 = keys[index+1]
        e2 = beg2end[b2] 
    
        # print("%s, %s, %s, %s" % (curr_beg, curr_end, b2, e2))
    
        if curr_beg <= b2 and curr_end >= e2: # do merge 
            if (index == len(keys)-2):  # have merged the last item, so, break 
                result.append([curr_beg, curr_end])
                break
        elif curr_beg <= b2 and curr_end >= b2: # do merge 
            curr_end = e2 
            if (index == len(keys)-2):  # have merged the last item, so, break 
                result.append([curr_beg, curr_end])
                break
        else: # cannot merge 
            result.append([curr_beg, curr_end])
            curr_beg = b2 
            curr_end = e2 
    
        index += 1
    
    return result 
    
    
result = merge([[1,3],[2,6],[8,10],[15,18]])
for item in result:
    print(item)
print("---------------------------")

result = merge([[1,4],[4,5]])
for item in result:
    print(item)
print("---------------------------")

result = merge([[1,4],[2, 3]])
for item in result:
    print(item)
print("---------------------------")

File: src/56_medium_MergeIntervals.cpp
class Solution {
public:
    vector<vector<int>> merge(vector<vector<int>>& intervals) {
        const int size = intervals.size();
        if (size <= 1) return intervals;

        sort(intervals.begin(), intervals.end(), 
             [](const vector<int>& a, const vector<int>& b){return a[0] < b[0];});
        
        vector<vector<int>> result;
        int p1 = 0, p2 = 1;
        int p1_left  = intervals[p1][0];
        int p1_right = intervals[p1][1];

        while (true) {
            if (p1_right < intervals[p2][0]) {  // cannot merge
                result.push_back({p1_left, p1_right});
                p1 = p2;
                p1_left  = intervals[p1][0];
                p1_right = intervals[p1][1];
                p2 ++;
                if (p2 >= size) {
                    result.push_back({p1_left, p1_right});
                    break;
                }
            } 
            else { // p1_right >= p2_left, so mergable
                p1_right = max(p1_right, intervals[p2][1]);
                if (p2 == size-1) {
                    result.push_back({p1_left, p1_right});
                    break;
                }
                else {
                    p2 ++;
                }
            }
        }
        return result;
    }
};